// @microsoft/teamsfx-core/build/common/spec-parser/interfaces.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidationStatus = exports.WarningType = exports.ErrorType = void 0;
/**
 * An enum that represents the types of errors that can occur during validation.
 */
var ErrorType;
(function (ErrorType) {
    ErrorType[ErrorType["SpecNotValid"] = 0] = "SpecNotValid";
    ErrorType[ErrorType["VersionNotSupported"] = 1] = "VersionNotSupported";
    ErrorType[ErrorType["RemoteRefNotSupported"] = 2] = "RemoteRefNotSupported";
    ErrorType[ErrorType["NoServerInformation"] = 3] = "NoServerInformation";
    ErrorType[ErrorType["MultipleServerInformation"] = 4] = "MultipleServerInformation";
    ErrorType[ErrorType["NoSupportedApi"] = 5] = "NoSupportedApi";
    ErrorType[ErrorType["ListFailed"] = 6] = "ListFailed";
    ErrorType[ErrorType["Cancelled"] = 7] = "Cancelled";
    ErrorType[ErrorType["Unknown"] = 8] = "Unknown";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
/**
 * An enum that represents the types of warnings that can occur during validation.
 */
var WarningType;
(function (WarningType) {
    WarningType[WarningType["AuthNotSupported"] = 0] = "AuthNotSupported";
    WarningType[WarningType["MethodNotSupported"] = 1] = "MethodNotSupported";
    WarningType[WarningType["OperationIdMissing"] = 2] = "OperationIdMissing";
    WarningType[WarningType["Unknown"] = 3] = "Unknown";
})(WarningType = exports.WarningType || (exports.WarningType = {}));
/**
 * An enum that represents the validation status of an OpenAPI specification file.
 */
var ValidationStatus;
(function (ValidationStatus) {
    ValidationStatus[ValidationStatus["Valid"] = 0] = "Valid";
    ValidationStatus[ValidationStatus["Warning"] = 1] = "Warning";
    ValidationStatus[ValidationStatus["Error"] = 2] = "Error";
})(ValidationStatus = exports.ValidationStatus || (exports.ValidationStatus = {}));
